src/aap_eda/api/serializers/activation.py

```diff
@@ -1005,93 +997,128 @@ def _get_aap_credentials_if_exists(
-        message = (
-            "No matching sources found for the following Event stream(s): "
-            f"{', '.join(sorted(missing_webhooks))}. {source_message}"
-        )
-        raise serializers.ValidationError({"webhooks": [message]})
-
-
-def _get_source_names(rulesets: list[dict]) -> set:
-    """Get all the source names associated with all rulesets."""
-    source_names = set()
-    for ruleset in rulesets:
-        for source in ruleset["sources"]:
-            if "name" in source:
-                source_names.add(source["name"])
-    return source_names
-
-
-def _get_source_count(rulesets: list[dict]) -> int:
-    """Get the number of sources in all rulesets."""
-    source_count = 0
-    for ruleset in rulesets:
-        source_count += len(ruleset["sources"])
-    return source_count
-
-
-def _get_webhook_names(webhook_ids: list[int]) -> set:
-    """Get all the webhook names associated with the ids."""
-    webhook_names = set()
-    for webhook_id in webhook_ids:
-        webhook = models.Webhook.objects.get(id=webhook_id)
-        if webhook:
-            webhook_names.add(webhook.name)
-    return webhook_names
-
-
-def _check_single_source_swap(rulesets: list[dict], webhook_names: set):
-    """Validate if a single source swap is possible."""
-    if _get_source_count(rulesets) != 1:
-        message = (
-            "You have more than 1 source in the rulebook. "
-            "Please add name to your sources and disable this option."
+    # validate no duplicate sources/webhooks in provided source mappings
+    for name in ["source_name", "webhook_name"]:
+        _check_duplicate_names(name, source_mappings)
+
+    # validate rulebook is not updated during mapping
+    rulebook_hash = get_rulebook_hash(rulebook.rulesets)
+    for source_map in source_mappings:
+        if source_map["rulebook_hash"] != rulebook_hash:
+            msg = (
+                "Rulebook has changed since the sources were mapped. "
+                "Please reattach Event streams again"
+            )
+
+            raise serializers.ValidationError(
+                {SOURCE_MAPPING_ERROR_KEY: [msg]}
+            )
+
+    # validate source_name provided in mapping exists in sources
+    source_names = [source["name"] for source in sources]
+    for source_map in source_mappings:
+        if source_map["source_name"] not in source_names:
+            msg = f"The source {source_map['source_name']} does not exist"
+            raise serializers.ValidationError(
+                {SOURCE_MAPPING_ERROR_KEY: [msg]}
+            )
+
+    # validate webhook ids and names
+    data["webhooks"] = _get_webhook_ids(source_mappings)
+
+
+def _validate_source_mappings(mappings: list[dict]) -> None:
+    for mapping in mappings:
+        missing_keys = [key for key in REQUIRED_KEYS if key not in mapping]
+
+    if missing_keys:
+        msg = (
+            f"The source mapping {mapping} missed the required keys: "
```

Comment:
@hsong-rh 
```
f"The source mapping {mapping} missed the required keys: "
```
maybe this would be better
```
  f"The source mapping is missing the required keys: "
```